sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -213,10 +213,51 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   object CartesianProduct extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
       case logical.Join(left, right, _, None) =>
-        execution.joins.CartesianProduct(planLater(left), planLater(right)) :: Nil
+        // For BroadcastCartesianProduct we will broadcast the small size plan,
+        // for CartesianProduct we will use the small size plan as cartesian left rdd.
+        if (right.statistics.sizeInBytes <= left.statistics.sizeInBytes) {
+          if (sqlContext.conf.autoBroadcastJoinThreshold > 0 &&
+            right.statistics.sizeInBytes <= sqlContext.conf.autoBroadcastJoinThreshold) {
```

Comment:
You can use ```CanBroadcast``` extractor in order to determine that a side is Broadcastable, this will also consider the ```broadcast``` hint, and is probably more future proof.